File: www/v2/adapter/MemoBaseAdapter.php
<?php

class MemoBaseAdapter extends DbAdapter implements SyncAdapter {

	const SYNC_TABLE = "MemoBases";
	
	public function get($memoBaseId) {
		return self::getDb(parent::connect(), $memoBaseId);
	}

	public static function getDb($db, $memoBaseId) {

		$stm = $db->prepare("SELECT * FROM memoling_MemoBases WHERE MemoBaseId = :MemoBaseId");
		$stm->bindParam(":MemoBaseId", $memoBaseId);
		$stm->execute();

		if($row = $stm->fetch()) {
			$obj = new MemoBase();
			$obj->Active = $row["Active"];
			$obj->Created = $row["Created"];
			$obj->MemoBaseId = $row["MemoBaseId"];
			$obj->Name = $row["Name"];
			return $obj;
		}

		return null;
	}

	public function getAll() {

		$db = parent::connect();
		$stm = $db->prepare("SELECT * FROM memoling_MemoBases");
		$stm->execute();

		$list = array();
		while($row = $stm->fetch()) {
			$obj = new MemoBase();
			$obj->Active = $row["Active"];
			$obj->Created = $row["Created"];
			$obj->MemoBaseId = $row["MemoBaseId"];
			$obj->Name = $row["Name"];
			$list[] = $obj;
		}

		return $list;

	}
    
    public function getAllForNavList($facebookUserId) {
        $db = parent::connect();
        
        $query = "SELECT 
            MB.MemoBaseId AS MemoBaseId, MB.Name AS Name,
            COUNT(M.MemoId) AS Count 
            FROM memoling_MemoBases AS MB
            LEFT OUTER JOIN memoling_Memos AS M 
                ON MB.MemoBaseId = M.MemoBaseId
            GROUP BY MB.MemoBaseId, MB.FacebookUserId
            HAVING MB.FacebookUserId = '100002530762250'
            ORDER BY MB.Name";
        
		$stm = $db->prepare($query);
        $stm->bindParam("FacebookUserId", $facebookUserId);
		$stm->execute();

		$list = array();
		while($row = $stm->fetch()) {
			$obj = new MemoBaseViewModel();
			$obj->Name = $row["Name"];
			$obj->Count = $row["Count"];
			$obj->MemoBaseId = $row["MemoBaseId"];
			$obj->Languages = self::getMemoLanguagesDb($db, $obj->MemoBaseId);
			$list[] = $obj;
		}
        
		return $list;
    }
	
    public static function getMemoLanguagesDb($db, $memoBaseId) {
        
        $query = "SELECT Languages FROM  (
                    SELECT 
                        DISTINCT WA.LanguageIso639 AS Languages
                    FROM memoling_Memos AS M
                    JOIN memoling_Words AS WA 
                        ON M.WordAId = WA.WordId
                    WHERE M.MemoBaseId = :MemoBaseId
                    ) AS A UNION (
                    SELECT 
                        DISTINCT WB.LanguageIso639 AS Languages
                    FROM memoling_Memos AS M
                    JOIN memoling_Words AS WB 
                        ON M.WordBId = WB.WordId
                    WHERE M.MemoBaseId = :MemoBaseId
                    )";
        
		$stm = $db->prepare($query);
        $stm->bindParam("MemoBaseId", $memoBaseId);
		$stm->execute();

		$list = array();
		while($row = $stm->fetch()) {
            $language = $row["Languages"];
			$list[] = $language;
		}

		return $list;
    }
    
	public function getMemoCount($memoBaseId) {
		return self::getMemoCountDb(parent::connect(), $memoBaseId);
	}

	public static function getMemoCountDb($db, $memoBaseId) {
		$query = "SELECT COUNT(1) AS MemoCount FROM memoling_Memos WHERE MemoBaseId = :MemoBaseId";
		$stm = $db->prepare($query);
		$stm->bindParam("MemoBaseId", $memoBaseId);
		$stm->execute();
		
		if($row = $stm->fetch()) {
			return $row["MemoCount"];
		} else {
			return null;
		}
	}

	public function insert($memoBase, $syncClientId=null) {
		return self::insertDb(parent::transConnect(), $memoBase, $syncClientId);
	}

	public static function insertDb($dbTran, $memoBase, $syncClientId=null) {
		$syncAction = new SyncAction();
		$syncAction->Action = SyncAction::ACTION_INSERT;
		$syncAction->PrimaryKey = $memoBase->MemoBaseId;
		$syncAction->Table = self::SYNC_TABLE;
		$syncAction->SyncClientId = $syncClientId;
		
		return self::insertDbSync($dbTran, $memoBase, $syncAction);
	}
	
	public function update($memoBase, $syncClientId=null) {
		return self::updateDb(parent::transConnect(), $memoBase, $syncClientId);
	}

	public static function updateDb($dbTran, $memoBase, $syncClientId=null) {

		$updateColumns = array();
			// This is the case when data is modified by user not sync mechanism
			// Thats why we need to see what have been changed/
		
        $dbMemoBase = self::getDb($dbTran, $memoBase->MemoBaseId);

        if($dbMemoBase->Name != $memoBase->Name) {
            $updateColumns[] = "Name";
        }
        if($dbMemoBase->Active != $memoBase->Active) {
            $updateColumns[] = "Active";
        }
		
		foreach($updateColumns as $updateColumn) {
		
			$syncAction = new SyncAction();
			$syncAction->Action = SyncAction::ACTION_UPDATE;
			$syncAction->PrimaryKey = $dbMemoBase->MemoBaseId;
			$syncAction->Table = self::SYNC_TABLE;
			$syncAction->SyncClientId = $syncClientId;
			$syncAction->UpdateColumn = $updateColumn;
		
			self::updateDbSync($dbTran, $memoBase, $syncAction);
		}
        
		if($memoBase->Memos != null) {
            foreach($memo as $memoBase->Memos) {
                MemoAdapter::updateDb($dbTran, $memo, $syncClientId);
            }
        }
	}

	public function delete($memoBaseId, $syncClientId=null) {
		return self::deleteDb(parent::transConnect(), $memoBaseId, $syncClientId);
	}

	public static function deleteDb($dbTran, $memoBaseId, $syncClientId=null) {
		$syncAction = new SyncAction();
		$syncAction->Action = SyncAction::ACTION_DELETE;
		$syncAction->PrimaryKey = $memoBaseId;
		$syncAction->Table = self::SYNC_TABLE;
		$syncAction->SyncClientId = $syncClientId;

		return self::deleteDbSync($dbTran, $memoBaseId, $syncAction);
	}
	
	public function decodeEntity($json) {
		$memoBase = new MemoBase();
		$memoBase->decode($json);
		return $memoBase;
	}
	
	public function getEntity($primaryKey) {
		$obj = $this->get($primaryKey);
		return $obj;
	}

	public function insertEntity($dbTran, $object, $syncAction) {
		self::insertDbSync($dbTran, $object, $syncAction);
	}
	
	public function updateEntity($dbTran, $object, $syncAction) {
		self::updateDbSync($dbTran, $object, $syncAction);
	}
	
	public function deleteEntity($dbTran, $primaryKey, $syncAction) {
		self::deleteDbSync($dbTran, $primaryKey, $syncAction);
	}
	
	private static function insertDbSync($dbTran, $memoBase, $syncAction) {
		$dbTran->beginTransaction();
		
        if($syncAction->SyncClientId != null &&
           $memoBase->FacebookUserId == null) {
            $syncClient = SyncClientAdapter::getDb($dbTran, $syncAction->SyncClientId);
            $memoBase->FacebookUserId = $syncClient->FacebookUserId;
        }
        
		$query = "INSERT INTO
				memoling_MemoBases(MemoBaseId,Name,Created,Active,FacebookUserId)
				VALUES(:Bid,:Name,:Created,:Active,:FacebookUserId)";
		$stm = $dbTran->prepare($query);
		
		$stm->bindParam("Bid", $memoBase->MemoBaseId);
		$stm->bindParam("Name", $memoBase->Name);
		$stm->bindParam("Created", $memoBase->Created);
		$stm->bindParam("Active", $memoBase->Active, PDO::PARAM_INT);
        $stm->bindParam("FacebookUserId", $memoBase->FacebookUserId);
		
		if(!$stm->execute()) {
			$dbTran->rollBack();
			throw new SqlException("Failed to create MemoBase", $dbTran);
		}
	
		if($memoBase->Memos != null) {
			foreach($memoBase->Memos as $memo) {
				MemoAdapter::insertDb($dbTran, $memo, $syncAction->SyncClientId);
			}
		}
	
		SyncActionAdapter::insertAction($dbTran, $syncAction);
	
		$dbTran->commit();
	}
	
	private static function updateDbSync($dbTran, $memoBase, $syncAction) {
		$dbTran->beginTransaction();
	
		$query = "UPDATE memoling_MemoBases SET ";
		$value = null;
		$type = null;
	
		switch($syncAction->UpdateColumn) {
			case "Name":
				$query .= "Name = :Value ";
				$value = $memoBase->Name;
				$type = PDO::PARAM_STR;
				break;
			case "Active":
				$query .= "Active = :Value ";
				$value = $memoBase->Active;
				$type = PDO::PARAM_INT;
				break;
		}
	
		$query .= "WHERE MemoBaseId = :MemoBaseId";
	
		$stm = $dbTran->prepare($query);
		$stm->bindParam("MemoBaseId", $memoBase->MemoBaseId);
		$stm->bindParam("Value", $value, $type);
	
		if(!$stm->execute()) {
			$dbTran->rollBack();
			throw new SqlException("Failed to update MemoBase", $dbTran);
		}
	
		SyncActionAdapter::updateAction($dbTran, $syncAction);
	
		$dbTran->commit();
	}
	
	private static function deleteDbSync($dbTran, $memoBaseId, $syncAction) {
		$dbTran->beginTransaction();
	

		$memoBase = self::getDb($dbTran, $memoBaseId);
		$memos = MemoAdapter::getAllDb($dbTran, $memoBaseId);
		
        if($memoBase != null) {
        
            foreach($memos as $memo) {
                MemoAdapter::deleteDb($dbTran, $memo->MemoId, null);
                SyncActionAdapter::removeActions($dbTran, MemoAdapter::SYNC_TABLE, $memo->MemoId);
            }

        }
		
		$query = "DELETE FROM
					memoling_MemoBases
					WHERE
					MemoBaseId = :MemoBaseId";
		
		$stm = $dbTran->prepare($query);
		$stm->bindParam(":MemoBaseId", $memoBaseId);
		if(!$stm->execute()) {
			$db->rollBack();
			throw new SqlException("Failed to delete MemoBase", $dbTran);
		}

		SyncActionAdapter::deleteAction($dbTran, $syncAction);
	
		$dbTran->commit();
	}

}

?>

File: android/Memoling/src/app/memoling/android/adapter/MemoBaseAdapter.java
package app.memoling.android.adapter;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashSet;
import java.util.UUID;
import java.util.List;

import org.json.JSONException;
import org.json.JSONObject;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;
import app.memoling.android.db.DatabaseHelper;
import app.memoling.android.db.SqliteAdapter;
import app.memoling.android.entity.Language;
import app.memoling.android.entity.Memo;
import app.memoling.android.entity.MemoBase;
import app.memoling.android.entity.MemoBaseInfo;
import app.memoling.android.entity.SyncAction;
import app.memoling.android.helper.DateHelper;
import app.memoling.android.sync.cloud.ISyncAdapter;
import app.memoling.android.sync.cloud.ISyncEntity;

public class MemoBaseAdapter extends SqliteAdapter implements ISyncAdapter {

	public final static String TableName = "MemoBases";


	public MemoBaseAdapter(Context context) {
		super(context);
	}

	public MemoBaseAdapter(Context context, boolean persistant) {
		super(context, persistant);
	}
	
	public static MemoBase bindMemoBase(Cursor cursor) {
		return MemoBaseAdapter.bindMemoBase(cursor, "");
	}
	
	public static MemoBase bindMemoBase(Cursor cursor, String prefix) {
		MemoBase memoBase = new MemoBase();
		memoBase.setActive(DatabaseHelper.getBoolean(cursor, prefix + "Active"));
		memoBase.setCreated(DatabaseHelper.getDate(cursor, prefix + "Created"));
		memoBase.setMemoBaseId(DatabaseHelper.getString(cursor, prefix + "MemoBaseId"));
		memoBase.setName(DatabaseHelper.getString(cursor, prefix + "Name"));
		
		return memoBase;
	}

	private static ContentValues createValues(MemoBase base) {
		ContentValues values = new ContentValues();
		values.put("MemoBaseId", base.getMemoBaseId());
		values.put("Name", base.getName());
		values.put("Created", DateHelper.toNormalizedString(base.getCreated()));
		values.put("Active", base.getActive());
		return values;
	}

	public MemoBase get(String memoBaseId) {
		try {
			return get(getDatabase(), memoBaseId);
		} finally {
			closeDatabase();
		}
	}

	public ArrayList<MemoBase> getAll() {
		try {
			return MemoBaseAdapter.getAll(getDatabase());
		} finally {
			closeDatabase();
		}
	}

	public void insert(MemoBase memoBase, String syncClientId) throws SQLiteException {
		try {
			MemoBaseAdapter.insert(getDatabase(), memoBase, syncClientId);
		} finally {
			closeDatabase();
		}
	}

	public void update(MemoBase memoBase, String syncClientId) throws SQLiteException {
		try {
			MemoBaseAdapter.update(getDatabase(), memoBase, syncClientId);
		} finally {
			closeDatabase();
		}
	}

	public void delete(String memoBaseId, String syncClientId) throws SQLiteException {
		try {
			MemoBaseAdapter.delete(getDatabase(), memoBaseId, syncClientId);
		} finally {
			closeDatabase();
		}
	}
	
	public static ArrayList<MemoBase> getAll(SQLiteDatabase db) {
		Cursor cursor = null;

		try {
			List<MemoBase> memoBases = new ArrayList<MemoBase>();

			String query = "SELECT "
					+ "	B.MemoBaseId B_MemoBaseId, B.Name B_Name, B.Created B_Created, B.Active B_Active "
					+ "FROM MemoBases  AS B";

			cursor = db.rawQuery(query, null);

			while (cursor.moveToNext()) {
				MemoBase memoBase = MemoBaseAdapter.bindMemoBase(cursor, "B_");
				memoBases.add(memoBase);
			}

			return memoBases;

		} finally {
			if (cursor != null && !cursor.isClosed()) {
				cursor.close();
			}
		}
	}

	public MemoBaseInfo getMemoBaseInfo(String memoBaseId) {
		SQLiteDatabase db = null;
		Cursor cursor = null;

		try {
			MemoBaseInfo memoBaseInfo = null;
			MemoBase memoBase = null;

			db = getDatabase();

			String query = "SELECT " 
					+ "B.MemoBaseId B_MemoBaseId, B.Name B_Name, B.Created B_Created, "
					+ "B.Active B_Active, COUNT(M.MemoBaseId) NoAll, SUM(M.Active) NoActive, "
					+ "MAX(M.LastReviewed) AS LastReviewed " 
					+ "FROM MemoBases  AS B "
					+ "OUTER LEFT JOIN Memos AS M ON M.MemoBaseId = B.MemoBaseId " 
					+ "WHERE B.MemoBaseId = ?";

			cursor = db.rawQuery(query, new String[] { memoBaseId });

			if (cursor.moveToFirst()) {
				memoBaseInfo = new MemoBaseInfo();

				memoBase = MemoBaseAdapter.bindMemoBase(cursor, "B_");
				
				memoBaseInfo.setMemoBase(memoBase);
				memoBaseInfo.setLastReviewed(DatabaseHelper.optDate(cursor, "LastReviewed", new Date()));
				memoBaseInfo.setNoActiveMemos(DatabaseHelper.optInt(cursor, "NoActive", 0));
				memoBaseInfo.setNoAllMemos(DatabaseHelper.optInt(cursor, "NoAll", 0));
			
				query = "SELECT LanguageIso639 FROM( " 
						+"SELECT DISTINCT coalesce(MB.MemoBaseId, MA.MemoBaseId) AS MemoBaseID, W.LanguageIso639 "
						+"FROM Words AS W "
						+"OUTER LEFT JOIN Memos AS MA ON MA.WordAId = W.WordId " 
						+"OUTER LEFT JOIN Memos AS MB ON MB.WordBId = W.WordId "
						+") "
						+"WHERE MemoBaseId = ?";
				
				cursor.close();
				cursor = db.rawQuery(query, new String[] { memoBaseId });
	
				HashSet<Language> languages = new HashSet<Language>();
				while (cursor.moveToNext()) {
					languages.add(Language.parse(DatabaseHelper.getString(cursor, "LanguageIso639")));
				} 
				Language[] arrayLanguages = new Language[languages.size()];
				arrayLanguages = languages.toArray(arrayLanguages);
				memoBaseInfo.setLanguages(arrayLanguages);
			}

			return memoBaseInfo;
		} finally {
			if (cursor != null && !cursor.isClosed()) {
				cursor.close();
			}

			closeDatabase();
		}
	}
	
	public static MemoBase get(SQLiteDatabase db, String memoBaseId) {
		
		String sql = "SELECT * FROM MemoBases WHERE MemoBaseId = ?";
		Cursor cursor = db.rawQuery(sql, new String[] { memoBaseId });
		
		try {
			
			if(cursor.moveToNext()) {
				MemoBase memoBase = MemoBaseAdapter.bindMemoBase(cursor);
				return memoBase;
			}
			
		} finally {
			cursor.close();
		}
		
		return null;
	}
	
	public static void insert(SQLiteDatabase db, MemoBase memoBase, String syncClientId) throws SQLiteException {
		SyncAction syncAction = new SyncAction();
		syncAction.setAction(SyncAction.ACTION_INSERT);
		syncAction.setPrimaryKey(memoBase.getMemoBaseId());
		syncAction.setTable(TableName);
		syncAction.setSyncClientId(syncClientId);
		
		MemoBaseAdapter.insertDbSync(db, memoBase, syncAction);
	}
	
	public static void update(SQLiteDatabase db, MemoBase memoBase, String syncClientId) throws SQLiteException {
		
		ArrayList<String> updateColumns = new ArrayList<String>();
		MemoBase dbMemoBase = MemoBaseAdapter.get(db, memoBase.getMemoBaseId());
		

		if(!dbMemoBase.getName().equals(memoBase.getName())) {
			updateColumns.add("Name");
		}
		
		if(dbMemoBase.getActive() == memoBase.getActive()) {
			updateColumns.add("Active");
		}
		
		for(String updateColumn : updateColumns) {
			SyncAction syncAction = new SyncAction();
			syncAction.setAction(SyncAction.ACTION_UPDATE);
			syncAction.setPrimaryKey(memoBase.getMemoBaseId());
			syncAction.setTable(TableName);
			syncAction.setSyncClientId(syncClientId);
			syncAction.setUpdateColumn(updateColumn);
			
			MemoBaseAdapter.updateDbSync(db, memoBase, syncAction);
		}
	}
	
	public static void delete(SQLiteDatabase db, String memoBaseId, String syncClientId) throws SQLiteException {
		SyncAction syncAction = new SyncAction();
		syncAction.setAction(SyncAction.ACTION_DELETE);
		syncAction.setPrimaryKey(memoBaseId);
		syncAction.setTable(TableName);
		syncAction.setSyncClientId(syncClientId);
		
		MemoBaseAdapter.deleteDbSync(db, memoBaseId, syncAction);
	}
	
	@Override
	public ISyncEntity decodeEntity(JSONObject json) throws JSONException {
		MemoBase memoBase = new MemoBase();
		memoBase.decodeEntity(json);		
		return memoBase;
	}

	@Override
	public ISyncEntity getEntity(String primaryKey) {
		return MemoBaseAdapter.get(this.getDatabase(), primaryKey);
	}

	@Override
	public void insertEntity(SQLiteDatabase db, ISyncEntity object, SyncAction syncAction) throws SQLiteException {
		MemoBaseAdapter.insertDbSync(db, (MemoBase)object, syncAction);
	}

	@Override
	public void deleteEntity(SQLiteDatabase db, String primaryKey, SyncAction syncAction) throws SQLiteException {
		MemoBaseAdapter.deleteDbSync(db, primaryKey, syncAction);
	}

	@Override
	public void updateEntity(SQLiteDatabase db, ISyncEntity object, SyncAction syncAction) throws SQLiteException {
		MemoBaseAdapter.updateDbSync(db, (MemoBase)object, syncAction);
	}
	@Override
	public void buildInitialSync(SQLiteDatabase db, String syncClientId) {
		replaceAllIdsDeep(db);
		
		String sql = "SELECT MemoBaseId FROM MemoBases";
		
		Cursor cursor = null;
		try {
			cursor = db.rawQuery(sql, null);
			while(cursor.moveToNext()) {
				SyncAction syncAction = new SyncAction();
				syncAction.setAction(SyncAction.ACTION_INSERT);
				syncAction.setTable(TableName);
				syncAction.setPrimaryKey(DatabaseHelper.getString(cursor, "MemoBaseId"));
				syncAction.setSyncClientId(syncClientId);
				
				SyncActionAdapter.insert(db, syncAction);
			}
		} finally {
			if(cursor != null) {
				cursor.close();
			}
		}
	}
	
	private static void insertDbSync(SQLiteDatabase db, MemoBase memoBase, SyncAction syncAction) throws SQLiteException {
		db.beginTransaction();

		try {
			ContentValues values = createValues(memoBase);
			
			if(memoBase.getMemos() != null) {
				for(Memo memo : memoBase.getMemos()) {
					MemoAdapter.insert(db, memo, syncAction.getSyncClientId());
				}
			}
			
			db.insertOrThrow(TableName, null, values);
			SyncActionAdapter.insertAction(db, syncAction);
			
			db.setTransactionSuccessful();
		} finally {
			db.endTransaction();
		}
	}
	
	private static void updateDbSync(SQLiteDatabase db, MemoBase memoBase, SyncAction syncAction) throws SQLiteException {
		db.beginTransaction();
		
		try {
			ContentValues values = new ContentValues();
			String column = syncAction.getUpdateColumn();
			
			if(column.equals("Name")) {
				values.put("Name", memoBase.getName());
			} else if(column.equals("Active")) {
				values.put("Active", memoBase.getActive());
			}

			if(values.size() > 0) {
				db.update(TableName, values, "MemoBaseId = ?", new String[] { memoBase.getMemoBaseId() });
				SyncActionAdapter.updateAction(db, syncAction);
			}
			
			db.setTransactionSuccessful();
		} finally {
			db.endTransaction();
		}
	}
	
	private static void deleteDbSync(SQLiteDatabase db, String memoBaseId, SyncAction syncAction) throws SQLiteException {
		db.beginTransaction();
		
		try {

			ArrayList<Memo> memos = MemoAdapter.getAll(db, memoBaseId);
			
			for(Memo memo : memos) {
				MemoAdapter.delete(db, memo.getMemoId(), null);
				SyncActionAdapter.removeAction(db, MemoAdapter.TableName, memo.getMemoId());
			}
			
			
			db.delete(TableName, "MemoBaseId = ?", new String[] { memoBaseId });
			SyncActionAdapter.deleteAction(db, syncAction);
			
			db.setTransactionSuccessful();
		} finally {
			db.endTransaction();
		}
	}
	
	private static void replaceAllIdsDeep(SQLiteDatabase db) throws SQLiteException {
		// This operation is part of initial sync. 
		// Do not record any sync actions
		
		db.beginTransaction();
		
		try {
			ArrayList<MemoBase> memoBases = MemoBaseAdapter.getAll(db);
			for(MemoBase memoBase : memoBases) {
				String oldMemoBaseId = memoBase.getMemoBaseId();
				String newMemoBaseId = UUID.randomUUID().toString();
				
				ArrayList<Memo> memos = MemoAdapter.getAll(db, oldMemoBaseId);
				for(Memo memo : memos) {
					String oldMemoId = memo.getMemoId();
					String newMemoId = UUID.randomUUID().toString();
	
					String oldWordAId = memo.getWordAId();
					String newWordAId = UUID.randomUUID().toString();
					
					String oldWordBId = memo.getWordBId();
					String newWordBId = UUID.randomUUID().toString();

					// Update words
					ContentValues vWordA = new ContentValues();
					vWordA.put("WordId", newWordAId);
					db.update(WordAdapter.TableName, vWordA, "WordId = ?", new String[] { oldWordAId });
					
					ContentValues vWordB = new ContentValues();
					vWordB.put("WordId", newWordBId);
					db.update(WordAdapter.TableName, vWordB, "WordId = ?", new String[] { oldWordBId });
					
					// Update Memo
					ContentValues vMemo = new ContentValues();
					vMemo.put("WordAId", newWordAId);
					vMemo.put("WordBId", newWordBId);
					vMemo.put("MemoId", newMemoId);
					vMemo.put("MemoBaseId", newMemoBaseId);
					
					db.update(MemoAdapter.TableName, vMemo, "MemoId = ?", new String[] { oldMemoId });
				}
				
				// Update MemoBase
				ContentValues vMemoBase = new ContentValues();
				vMemoBase.put("MemoBaseId", newMemoBaseId);
				db.update(MemoBaseAdapter.TableName, vMemoBase, "MemoBaseId = ?", new String[] { oldMemoBaseId });
			}
		
			db.setTransactionSuccessful();
		} finally {
			db.endTransaction();
		}
	}
}
